xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -0,0 +1,479 @@
+     */
+    private double virtualTimeNow = 0.0;
+
+    /**
+     * Weights below this value will be logged and upped to this minimum weight.
+     */
+    private static final double MINIMUM_WEIGHT = 0.0001;
+
+    private final Object lock = new Object();
+
+    /**
+     * Use the item's deadline as the order in the priority queue. If the deadlines are the same,
+     * use the index. Index should be unique.
+     */
+    EdfScheduler() {
+      this.prioQueue = new PriorityQueue<ObjectState>(10, (o1, o2) -> {
+        if (o1.deadline == o2.deadline) {
+          return o1.index - o2.index;
+        } else if (o1.deadline < o2.deadline) {
+          return -1;
+        } else {
+          return 1;
+        }
+      });
+    }
+
+    /**
+     * Adds (or updates) the item in the scheduler. This is not thread safe.
+     *
+     * @param index The field {@link ObjectState#index} to be added/updated
+     * @param weight positive weight for the added/updated object
+     */
+    void add(int index, double weight) {
+      checkArgument(weight > 0.0, "Weights need to be positive.");
+      ObjectState state = new ObjectState(Math.max(weight, MINIMUM_WEIGHT), index);
+      state.deadline = virtualTimeNow + 1 / state.weight;
+      prioQueue.add(state);
+    }
+
+    /**
+     * Picks the next WRR object.
+     */
+    int pick() {
+      synchronized (lock) {
+        ObjectState minObject = prioQueue.remove();
+        // Simulate advancing in time by setting the current time to the period of the nearest item
+        // on the "time horizon".
+        virtualTimeNow = minObject.deadline;
+        minObject.deadline = virtualTimeNow + (1.0 / minObject.weight);
+        prioQueue.add(minObject);
+        return minObject.index;
+      }
+    }
+  }
+
+  /** Holds the state of the object. */
+  @VisibleForTesting
+  static class ObjectState {
+    private final double weight;
+    private final int index;
+    volatile double deadline;
+
+    ObjectState(double weight, int index) {
+      this.weight = weight;
+      this.index = index;
+    }
+  }
+
+  static final class WeightedRoundRobinLoadBalancerConfig {
+    final Long blackoutPeriodNanos;
+    final Long weightExpirationPeriodNanos;
+    final Boolean enableOobLoadReport;
+    final Long oobReportingPeriodNanos;
+    final Long weightUpdatePeriodNanos;
+
+    public static Builder newBuilder() {
+      return new Builder();
+    }
+
+    private WeightedRoundRobinLoadBalancerConfig(Long blackoutPeriodNanos,
+                                                 Long weightExpirationPeriodNanos,
+                                                 Boolean enableOobLoadReport,
+                                                 Long oobReportingPeriodNanos,
+                                                 Long weightUpdatePeriodNanos) {
+      this.blackoutPeriodNanos = blackoutPeriodNanos;
+      this.weightExpirationPeriodNanos = weightExpirationPeriodNanos;
```

Comment:
Good call. 
This is something that needs cross-language consistency. The design didn't include it. Its most likely misconfiguration and might be fair to be called as user's responsibility tho. Maybe we can add as a future improvement. cc. @markdroth  @yousukseung

Reply:
They are two separate things.

Weight expiration kicks in if the backend does not report load metrics for too long (i.e. backend report -> weight in subchannel), and weight_update_period determines how often we update the scheduler (weight in subchannel -> scheduler).

Does the current implementation do otherwise?

Reply:
A quick check shows the implementation is consistent with C++ so I think we're good.

Reply:
If weightExpirationPeriodNanos < weightUpdatePeriodNanos then you are going to have a period of time every cycle that you have no weight.  For example if we expire in 2 seconds and update every minute, then for 58 seconds of each minute we think the weight is unusable.

Reply:
No, they are orthogonal. The current implementation already does it right so I think you can check code.

If expiration=2s, then we'll get a zero weight (=use average) only from subchannels that did not receive a load report for 2s or longer at the time GetWeight() is called. Obviously we expect load reports to come more frequently than the expiration time configured. The weight update period decides how often we re-create the scheduler, in which GetWeight() is called to all subchannels.

Reply:
Thanks! I see. `getWeight()` that is periodically by `weightUpdatePeriodNanos` indeed does not effect the expiration.  They control separate things about the metrics: frequency and duration, same between expiration configuration and OOB request period configuration.  Typically effective weights are coming more often than the expiration, but it might still be fine if not.

Reply:
My point is that on average in my scenario that most of the time you will be getting 0 weight.  While they perform different functions, there is a requirement that the results are relatively close to proportional to the weights.  If most of the time the weights are 0, then you aren't really weighting properly.  If you have 2 subchannels with weights 59 and 1, you would expect that 12000 evenly spaced requests over a minute would be split 11800 : 200.  However what really happens is that 400 are split ~ 393 : 7 and the rest are split 5800 : 5800 so the total is 6193 : 5807 which is nothing like what is desired.

Reply:
Note that I don't think this discussion should block this PR, it is more about having follow on PRs in all languages to protect the users from misconfiguring.